### PR TITLE
Add BigInt

### PR DESCRIPTION
**Pull request by adrianheine**
This is my try at representing what I understood to be the agreed-upon solution in #169.

### CODE REVIEW

File: experimental/bigint.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# [BigInt](https://github.com/tc39/proposal-bigint)
+
+```js
+interface BigIntLiteral <: Literal {
+  bigint: string;
```
Comment (michaelficarra) [inline experimental/bigint.md]: Other `Literal`s use `value` as the property name.

Comment (RReverser) [reply experimental/bigint.md]: As far as I understand, the idea here is to have it similar to `RegExpLiteral` where `value` might or might not be representable on the target runtime so another field is needed to represent the parsing result in these cases.

Comment (RReverser) [reply experimental/bigint.md]: Actually, you were the one who suggested that in https://github.com/estree/estree/issues/169#issuecomment-329059842

Comment (michaelficarra) [reply experimental/bigint.md]: Ah, that's right, there's no `"type"` or anything distinguishing it from other `Literal`s. Got it. Okay, what should the `value` of these nodes be? Currently bigints are not allowed in the `value` field.

Comment (adrianheine) [reply experimental/bigint.md]: You suggested to use a BigInt literal for `value`. For that we would need to amend the `Literal` interface (`extend interface Literal { type: string | boolean | null | number | RegExp | bigint;`) and add `value: bigint` to this interface, right?

Comment (michaelficarra) [reply experimental/bigint.md]: Yep, that should do it.

Comment (adrianheine) [reply experimental/bigint.md]: I pushed it, but with property `value`, not `type`.

File: experimental/bigint.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# [BigInt](https://github.com/tc39/proposal-bigint)
+
+```js
+extend interface Literal {
+  value: string | boolean | null | number | RegExp | bigint;
+}
+
+interface BigIntLiteral <: Literal {
+  value: bigint;
+  bigint: string;
```
Comment (j-f1) [inline experimental/bigint.md]: How about `raw` and `cooked`? (i.e.`raw` would be `"12345n"` and cooked would be `12345n` (as a `BigInt`)?

Comment (adrianheine) [reply experimental/bigint.md]: `cooked` only exists with template literals. `raw` is not part of the spec for (non-template) literals (see #14).

As for platforms without BigInt support, I've been asking this myself, too. Babylon doesn't emit the value, so I'm not aware of any precedents. I can think of a) emitting null b) casting to a regular number, which is lossy.

Comment (j-f1) [reply experimental/bigint.md]: FYI “platforms without `BigInt` support” includes (and will include, for all eternity<sup>[*citation needed*]</sup>) JSON.

Comment (adrianheine) [reply experimental/bigint.md]: You are right, just as regular expressions. There's a discussion in #64 about JSON.

Comment (RReverser) [reply experimental/bigint.md]: I think emitting `null` if bigint is not in safe integer space makes sense and would be in line with regular expression literals.

Comment (michaelficarra) [reply experimental/bigint.md]: Why safe integer space? Use a bigint if the platform supports it (regardless of magnitude), `null` otherwise.

Comment (adrianheine) [reply experimental/bigint.md]: I've made `value: bigint | null` for now, but I would still prefer to drop or phase out `value` as explained in https://github.com/estree/estree/issues/64#issuecomment-357434955.

**Comment by goodmind** [comment]
Should it be null if integer isn't in safe space or always null? (for JSON dump of AST)

**Pull request closed without merging**